### ISSUE TITLE
RDFParser throws an error "Not a valid token for an RDF term: [LBRACE]" while parsing valid Turtle file

### ISSUE THREAD

**Issue by mpro7** [apache/jena#2083]
### Version

4.9.0

### What happened?

After upgrading Jena from `v4.8.0` to `v4.9.0` one of tests which process streamed RDF from [valid TTL file](https://github.com/dasch-swiss/dsp-api/blob/main/test_data/project_ontologies/anything-onto.ttl) and write the formatted result to an output stream failed with an error:
**sbt.ForkMain$ForkError: org.apache.jena.riot.RiotException: [line: 11, col: 26] Not a valid token for an RDF term: [LBRACE]**.

I've noticed that `v4.9.0` introduced new Turtle format, so maybe btw of that change a bug sneaked into the RDF parser code too?

### Relevant output and stacktrace

```shell
[line: 11, col: 26] Not a valid token for an RDF term: [LBRACE]
org.apache.jena.riot.RiotException: [line: 11, col: 26] Not a valid token for an RDF term: [LBRACE]
	at org.apache.jena.riot.system.ErrorHandlerFactory$ErrorHandlerStrict.fatal(ErrorHandlerFactory.java:235)
	at org.apache.jena.riot.system.ParserProfileStd.create(ParserProfileStd.java:310)
	at org.apache.jena.riot.lang.LangTurtleBase.tokenAsNode(LangTurtleBase.java:577)
	at org.apache.jena.riot.lang.LangTurtleBase.node(LangTurtleBase.java:410)
	at org.apache.jena.riot.lang.LangTurtleBase.predicate(LangTurtleBase.java:376)
	at org.apache.jena.riot.lang.LangTurtleBase.predicateObjectItem(LangTurtleBase.java:350)
	at org.apache.jena.riot.lang.LangTurtleBase.predicateObjectList(LangTurtleBase.java:333)
	at org.apache.jena.riot.lang.LangTurtleBase.triples(LangTurtleBase.java:314)
	at org.apache.jena.riot.lang.LangTurtleBase.triplesSameSubject(LangTurtleBase.java:178)
	at org.apache.jena.riot.lang.LangTurtle.oneTopLevelElement(LangTurtle.java:46)
	at org.apache.jena.riot.lang.LangTurtleBase.runParser(LangTurtleBase.java:79)
	at org.apache.jena.riot.lang.LangBase.parse(LangBase.java:43)
	at org.apache.jena.riot.RDFParserRegistry$ReaderRIOTLang.read(RDFParserRegistry.java:204)
	at org.apache.jena.riot.RDFParser.read(RDFParser.java:416)
	at org.apache.jena.riot.RDFParser.parseNotUri(RDFParser.java:406)
	at org.apache.jena.riot.RDFParser.parse(RDFParser.java:356)
	at org.apache.jena.riot.RDFParserBuilder.parse(RDFParserBuilder.java:570)
```

### Are you interested in making a pull request?

None

**Comment by rvesse** [comment]
I'm unable to reproduce this, the given input file validates successfully `riot --validate` with Jena 4.8.0, 4.9.0, 4.10.0

Also the error stack trace you show refers to a Line/Column that is a blank line in your linked file so that isn't the actual data you are encountering the error on.  Your bug report implies that there is some data transformation happening in your environment so you need to provide all the steps necessary to reproduce this issue in order for us to help you.

**Comment by mpro7** [comment]
@rvesse thanks for the replay. Sorry, the stacktrace was somehow truncated. [Here is the link](https://github.com/dasch-swiss/dsp-api/pull/2911/files#annotation_15514320445) how it exactly looks like (the bottom error). In the test the file is read and process with `parseWithStreamProcessor` method which uses `jena.riot.RDFParser` to do so.

**Comment by rvesse** [comment]
> [Here is the link](https://github.com/dasch-swiss/dsp-api/pull/2911/files#annotation_15514320445)

Again no idea what that link is for, it's to a diff in one of your projects PRs and there's no error message present there (nor any Jena related code) 😕

**Comment by mpro7** [comment]
@rvesse sorry, this is because I've updated the branch and instead of commit link I've pasted some failing tests link, which seems to change after the branch update. [Here is the correct link](https://github.com/dasch-swiss/dsp-api/pull/2911/commits/1058ee6777e4def029d5790559735a6afb47460b). To see the full stack trace, click raw output.

**Comment by afs** [comment]
> I've noticed that v4.9.0 introduced new Turtle format,

It is a new output format and it isn't used by default.

I can't see what's happening other than it seems the parsing is from an input stream given the parser. So setup at line 139.

There isn't an LBRACE (a `{` ) in the file.

This would be the error if the input stream is TriG syntax and the parser syntax is set to Turtle.

Please could we have a complete, reproducible, minimal example.

**Comment by afs** [comment]
@mpro7 Any progress on a reproducible, minimal example?

**Comment by afs** [comment]
We haven't heard back on this and can't reproduce it.

@mpro7 Please do reopen this issue if there is new information.

**Comment by mpro7** [comment]
@afs sorry for coming back late, we solved the issue which was on our side. Our code did some additional transformations which started to fail after upgrading to Jena v4.9.0. Once we cleaned up the code and started calling Jena functions more directly, the problem was gone. However it shows that some functionalities on Jena side changed.

If you still would like to have more detailed information, please let me know.

**Comment by afs** [comment]
@mpro7 - Yes, I'd like to know more. 

Very little has changed between 4.8.0 (several versions before that) to 4.9.0 in this area.

**Comment by BalduinLandolt** [comment]
Hi, since I worked with @mpro7 on this, and I had some spare time to reproduce it, let me provide additional information:

The following (admittedly nonsensical!) Scala code:

```scala
import org.apache.jena

import java.io.{ByteArrayInputStream, ByteArrayOutputStream, InputStream, OutputStream}
import java.nio.charset.StandardCharsets
import scala.util.Try

object Run extends App {
  val ttl =
    """|
       |@prefix xsd:        <http://www.w3.org/2001/XMLSchema#> .
       |@prefix rdf:        <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
       |@prefix rdfs:       <http://www.w3.org/2000/01/rdf-schema#> .
       |@prefix owl:        <http://www.w3.org/2002/07/owl#> .
       |@prefix knora-base: <http://www.knora.org/ontology/knora-base#> .
       |
       |<http://www.knora.org/ontology/0001/anything>
       |    rdf:type                        owl:Ontology ;
       |    rdfs:label                      "The anything ontology" ;
       |    knora-base:attachedToProject    <http://rdfh.ch/projects/0001> ;
       |    knora-base:lastModificationDate "2017-12-19T15:23:42.166Z"^^xsd:dateTime .
       |
       |""".stripMargin

  val is = new ByteArrayInputStream(ttl.getBytes(StandardCharsets.UTF_8))
  val os = new ByteArrayOutputStream()

  StreamParser.parse(is, os)
  val res = os.toString

  println(res)
}

object StreamParser {

  def parse(inputStream: InputStream, outputStream: OutputStream): Unit = {

    val streamRDF = new jena.riot.system.StreamRDF {
      val inner = jena.riot.system.StreamRDFWriter.getWriterStream(outputStream, jena.riot.RDFLanguages.TURTLE)

      override def start(): Unit                             = inner.start()
      override def finish(): Unit                            = inner.finish()
      override def base(base: String): Unit                  = {}
      override def prefix(prefix: String, iri: String): Unit = inner.prefix(prefix, iri)
      override def quad(quad: jena.sparql.core.Quad): Unit   = inner.quad(quad)
      // ----- !!! here !!! -----
      override def triple(triple: jena.graph.Triple): Unit =
        inner.quad(jena.sparql.core.Quad.create(jena.sparql.core.Quad.defaultGraphIRI, triple))
    }
    val parser = jena.riot.RDFParser.create()
    parser.source(inputStream)

    val parseTry: Try[Unit] = Try {
      parser
        .lang(jena.riot.RDFLanguages.TURTLE)
        .errorHandler(jena.riot.system.ErrorHandlerFactory.errorHandlerStrictNoLogging)
        .parse(streamRDF)
    }
    inputStream.close()
    outputStream.close()
    parseTry.get
  }

}

```

when executed with Jena 4.8.0, produces

```
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix knora-base: <http://www.knora.org/ontology/knora-base#> .

<http://www.knora.org/ontology/0001/anything>
        rdf:type                      owl:Ontology ;
        rdfs:label                    "The anything ontology" ;
        knora-base:attachedToProject  <http://rdfh.ch/projects/0001> ;
        knora-base:lastModificationDate  "2017-12-19T15:23:42.166Z"^^xsd:dateTime .

```

but with Jena 4.9.0 it outputs

```
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix knora-base: <http://www.knora.org/ontology/knora-base#> .

<urn:x-arq:DefaultGraph> {
    <http://www.knora.org/ontology/0001/anything>
            rdf:type                      owl:Ontology ;
            rdfs:label                    "The anything ontology" ;
            knora-base:attachedToProject  <http://rdfh.ch/projects/0001> ;
            knora-base:lastModificationDate  "2017-12-19T15:23:42.166Z"^^xsd:dateTime .
}

```

the interesting part is probably where the implementation of `StreamRDF` implements `triple()` as `quad()` with default graph.  
From the looks of it, in Jena 4.8 this produced TTL, while in Jena 4.9 it seems to produce TRIG output.

I hope this information helps, and I apologise for providing Scala rather than Java code (but it was painful enough to get to this reproduction of the issue).

If you have any questions, please feel free to ask!

**Comment by afs** [comment]
@BalduinLandolt -- thank you for the information.

```scala
      override def triple(triple: jena.graph.Triple): Unit =
        inner.quad(jena.sparql.core.Quad.create(jena.sparql.core.Quad.defaultGraphIRI, triple))
```

Could I ask what the code is trying to achieve by converting a triple to a quad if the intended output is Turtle (triples only)?

**Comment by BalduinLandolt** [comment]
> Could I ask what the code is trying to achieve by converting a triple to a quad if the intended output is Turtle (triples only)?

Not really, sorry... we're dealing with some rather gnarly legacy code, and in this part of the code base there were tons of abstractions and wrappers, because historically this was designed to work both with a Jena or an RDF4J model underneath. But RDF4J had been removed long ago, so the abstractions had become rather pointless. When I inlined all that was going on, I ended up with something similar to the sample code I pasted above - and that's when it became obvious that the implementation does not make sense.  
We ended up removing some of the redundant abstractions, calling functionality on the Jena models directly, and that had already fixed our problem (as @mpro7 had pointed out).  
In the real code base, the function would also take in an RDFLanguage as a param; and then it would be used to transform a stram of RDF statements from one language to another - but that doesn't make `inner.quad(jena.sparql.core.Quad.create(jena.sparql.core.Quad.defaultGraphIRI, triple))` more sensible, of course.

In any case, our issue is fixed, no need to worry about that. But my example proves that the behaviour from Jena 4.8 to 4.9 has changed somewhat (and arguable a TTL-Parser should probably not return TRIG, even if it gets quads for triples), so if you want to investigate that, I hope my example helps.
